tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -177,6 +178,106 @@ DECL_CONVERT_OP(FakeQuant);
 
 #undef DECL_CONVERT_OP
 
+LogicalResult ConvertTFLGeluOp::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_gelu_op = cast<TFL::GeluOp>(op);
+  Location loc = op->getLoc();
+
+  Value input = tfl_gelu_op.input();
+  RankedTensorType input_type = input.getType().dyn_cast<RankedTensorType>();
+  RankedTensorType output_type =
+      tfl_gelu_op.getResult().getType().dyn_cast<RankedTensorType>();
+  if (!input_type || !output_type) {
+    return rewriter.notifyMatchFailure(
+        op, "input/output are not all a ranked tensor");
+  }
+
+  UniformQuantizedType in_quant_type =
+      input_type.getElementType().dyn_cast<mlir::quant::UniformQuantizedType>();
+  UniformQuantizedType out_quant_type =
+      output_type.getElementType()
+          .dyn_cast<mlir::quant::UniformQuantizedType>();
+
+  if ((in_quant_type == nullptr) != (out_quant_type == nullptr)) {
+    return rewriter.notifyMatchFailure(
+        op,
+        "input/output tensor should be all quantized or all floating-point");
+  }
+
+  if (out_quant_type) {
+    // The formal definition of gelu.
+    auto gelu_func = [](double x) -> double {
+      return 0.5 * x * (1.0 + std::erf(x / std::sqrt(2)));
+    };
+
+    if (in_quant_type.getStorageTypeIntegralWidth() != 8) {
+      return rewriter.notifyMatchFailure(
+          op, "current tfl.gelu only support 8-bit quantized type");
+    }
+
+    Value table_const = getTosaConst8bitTable(
+        rewriter, op, in_quant_type.getScale(), in_quant_type.getZeroPoint(),
+        out_quant_type.getScale(), out_quant_type.getZeroPoint(), gelu_func);
+
+    CreateReplaceOpAndInfer<tosa::TableOp>(rewriter, op, output_type, input,
+                                           table_const);
+    return success();
+  }
+
+  // Following approximated implemention described in
+  // tensorflow/lite/kernels/internal/reference/gelu.h
+  // gelu(x) = 0.5 * x * (1 + tanh(sqrt(2/pi) * (x + 0.044715 * x^3)))
+  // Lower the formula to the sequence of operators below.
+  // op0 = pow(x, 3)
+  // op1 = mul(op0, 0.044715)
+  // op2 = add(x, op1)
+  // op3 = mul(op2, sqrt(2/pi))
+  // op4 = tanh(op3)
+  // op5 = add(op4 ,1)
+  // op6 = mul(x, 0.5)
+  // op7 = mul(op6, op5)
+
+  auto fp_scalar_ty = RankedTensorType::get({}, rewriter.getF32Type());
+
+  Value cst_3 = rewriter.create<tosa::ConstOp>(
+      loc, fp_scalar_ty, DenseElementsAttr::get(fp_scalar_ty, {3.000000e+00f}));
+  auto op0_pow_3 =
```

Comment:
This does not need to be addressed here, but these really make me feel like we should have a nicer builder for these ... I feel like the structure is being obstructed. The numbering in variables also make these a bit fragile to changes (e.g., reordering two instructions above that doesn't result computation chain resulting in needing to update more in the file)

Reply:
Agree. IMO, there is a probably way to simplify this by extending `CreateOpAndInfer` to accept the numeric constant argument. Create `Value` from the data type and value of constant parameter passed in.